### PR TITLE
Migration to FindPython3.cmake

### PR DESCRIPTION
**Pull request by ilya-lavrenov**
### Description
- Migration to `FindPython3.cmake` as this module is new cmake approach and contains many fixes compared with old ones. E.g. it can search  python from activated virtual envs automatically 
- Removed non-necessary variables specifying like `PYTHON_INCLUDE_DIR` which are auto discovered by new module
- Used cmake 3.18 as it contains stable version of `FindPython3.cmake` and `Development.Module` is available (see https://cmake.org/cmake/help/latest/module/FindPython3.html#module:FindPython3), which requires less dependencies to be installed on the system to build python extensions.
- Updated minimal pybind11 version to more stable one which consists fixes useful for cross-compilation
- Deprecated `BUILD_ONNX_PYTHON` and introduced `ONNX_BUILD_PYTHON` option to be aligned with all other cmake options to have a common `ONNX_` prefix.

### Motivation and Context
To prevent warning in new cmake versions that `find_package(PythonInterp)` and `find_package(PythonLibs)` are deprecated and will be dropped soon. See https://cmake.org/cmake/help/latest/module/FindPythonInterp.html - these modules are available only with special policy.

### CODE REVIEW

**Comment by justinchuby** [comment]
Looks like we have some build errors

**Review by cyyever — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
Fixed issue https://github.com/onnx/onnx/actions/runs/11656555709/job/32452821839?pr=6526 with latest push

File: CMakeLists.txt [hunk 1]
```diff
@@ -527,9 +543,9 @@ target_include_directories(onnx PUBLIC
 target_link_libraries(onnx PUBLIC onnx_proto)
 add_onnx_global_defines(onnx)
 
-if(BUILD_ONNX_PYTHON)
-  if("${PY_EXT_SUFFIX}" STREQUAL "")
-    if(MSVC)
+if(ONNX_BUILD_PYTHON)
+  if(NOT PY_EXT_SUFFIX)
+    if(WIN32)
```
Comment (justinchuby) [inline CMakeLists.txt:548]: Could you share the reason for changing msvc to win32? Thanks - just for my education

**Review by justinchuby — APPROVED** [review]
LGTM thanks

**Comment by justinchuby** [comment]
cc @gramalingam or @liqunfu for another review

**Comment by justinchuby** [comment]
Could you resolve the merge conflicts? Thanks

**Comment by justinchuby** [comment]
Also mention the option deprecation introduced in the PR in the pr description

**Pull request merged**